### ISSUE TITLE
Issues with Streaming API Responses to Frontend on AWS

### ISSUE THREAD

**Issue by anni1236012** [vercel/ai#635]
### Description

Streaming is working fine locally but after deploying it on AWS amplify it does not work. 

### Code example

**API Route**

```javascript
import { StreamingTextResponse, LangChainStream, Message } from 'ai';
import { ChatOpenAI } from 'langchain/chat_models/openai';
import { AIMessage, HumanMessage } from 'langchain/schema';

export const runtime = 'edge';

export async function POST(req: Request) {
  const { messages } = await req.json();

  const { stream, handlers } = LangChainStream();

  const llm = new ChatOpenAI({
    streaming: true,
  });

  llm
    .call(
      (messages as Message[]).map(m =>
        m.role == 'user'
          ? new HumanMessage(m.content)
          : new AIMessage(m.content),
      ),
      {},
      [handlers],
    )
    .catch(console.error);

  return new StreamingTextResponse(stream);
}
```

UI

```javascript
'use client';

import { useChat } from 'ai/react';

export default function Chat() {
  const { messages, input, handleInputChange, handleSubmit } = useChat();

  return (
    <div className="mx-auto w-full max-w-md py-24 flex flex-col stretch">
      {messages.length > 0
        ? messages.map(m => (
            <div key={m.id} className="whitespace-pre-wrap">
              {m.role === 'user' ? 'User: ' : 'AI: '}
              {m.content}
            </div>
          ))
        : null}

      <form onSubmit={handleSubmit}>
        <input
          className="fixed w-full max-w-md bottom-0 border border-gray-300 rounded mb-8 shadow-xl p-2"
          value={input}
          placeholder="Say something..."
          onChange={handleInputChange}
        />
      </form>
    </div>
  );
}

```

**Comment by MaxLeiter** [comment]
I'm not sure AWS Amplify supports streaming (although their lamdas generally do so I think it should). You can read about how Vercel accomplishes it here: https://vercel.com/blog/streaming-for-serverless-node-js-and-edge-runtimes-with-vercel-functions

Did this work for you previously?

**Comment by anni1236012** [comment]
No, I am trying this for the first time. Lambda function support streaming so let me try moving nextJs API route to lambda function and call it directly from the client side. I will update it here if that works.

**Comment by peterdresslar** [comment]
[It looks like] (https://aws.amazon.com/blogs/compute/introducing-aws-lambda-response-streaming/) you would need to wrap the stream logic into Amazon's special streaming decorator and deploy to lambda. Something to consider is that the documentation referenced, especially the 2nd code sample, is using [WriteableStream](https://nodejs.org/docs/latest-v14.x/api/stream.html#stream_class_stream_writable); the streams in `core/streams` here are `ReadableStream`. It seems possible you'd need to take a step to transform into something decoratable for the Lambda. A question like "how can I make a `ReadableStream` stream on Lambda in an AWS support forum might really help.

**Comment by anni1236012** [comment]
I've moved my Next.js API route to an AWS Lambda function URL which fixed a streaming issue. However, I'm now having trouble integrating it with the useChat hook due to the necessity of signing the request with AWS SigV4.

Here’s a brief of my issue:

I know headers can be passed to useChat.
To sign the request, I need access to the complete messages array, but the latest user question is appended to messages array inside useChat after handleSubmit is called. Some additonal information like createdAt and chatid is also added to messages array inside useChat. 
![image](https://github.com/vercel/ai/assets/86418669/db5ae97b-e842-4ed3-9290-303a90526468)

Is there a way to effectively sign requests to Lambda function URLs using SigV4 using useChat hook, considering the need to access the messages array post-handleSubmit? Any guidance on achieving this would be immensely appreciated.

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **None**  
Confidence: **High**

### Rationale
The reported streaming problem was attributable to the AWS deployment environment, and the follow-up is a usage question about AWS request signing rather than a demonstrated SDK failure or explicit enhancement request.

### Evidence
- Streaming worked locally and began working after moving the API route from AWS Amplify to an AWS Lambda function URL.
- Comments indicate AWS Amplify may not support the required response streaming behavior and identify no concrete AI SDK defect.
- The remaining request asks how to sign useChat requests with AWS SigV4, which is an integration/support question.
